src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -771,6 +771,13 @@ CHIP_ERROR TimeSynchronizationServer::GetDefaultNtp(MutableCharSpan & dntp)
 Span<TimeSyncDataProvider::TimeZoneStore> & TimeSynchronizationServer::GetTimeZone()
 {
     mTimeZoneObj.timeZoneList = mTimeZoneObj.timeZoneList.SubSpan(0, mTimeZoneObj.validSize);
+
+    // Make sure the sizes of the entries is correct
+    for (auto & tzStore : mTimeZoneObj.timeZoneList)
+    {
+        tzStore.timeZone.name = MakeOptional(CharSpan(tzStore.name, strnlen(tzStore.name, sizeof(tzStore.name))));
```

Comment:
not totally sure I understand this - the name is a char span, right? It should be backed by the char array in the TimeZoneStore, but its' still a span. I think this might be mis-initialized through. There are too many things called "name". But check line 554 in InitTimeZone. I think it's going in weird, not coming out weird.

Reply:
Line 554 makes it sizeof(name), so it makes a charspan of a 0-filled value. Unfortunately TLV writer will write it as-is, so you get ` Record { tag: ContextSpecific { tag: 2 }, value: Utf8([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) }` inside the TLV (see #30342).

Ideally this would be adjusted when we write, but this file is a 1.3K line code with mirrored data (why do we mirror the span? We should have only created those structures for the serialization code) so I did not try to fix it, just patch it to at least not write 64 bytes when not needed.

Reply:
... so I think it starts off weird, however I don't think we should care. If the backing is a c_str, that is fine, but when we serialize and we use charspan, it should be corrected to use the c_str length.

The bug seems to be that we have timeZoneList alltogether, however fixing that requires me to understand the logic a lot more.

Reply:
I think the real issue is:

```
void TimeSynchronizationServer::InitTimeZone()
{
    mTimeZoneObj.validSize    = 1; // one default time zone item is needed
    mTimeZoneObj.timeZoneList = Span<TimeSyncDataProvider::TimeZoneStore>(mTz);
    for (auto & tzStore : mTimeZoneObj.timeZoneList)
    {
        memset(tzStore.name, 0, sizeof(tzStore.name));
        tzStore.timeZone = { .offset = 0, .validAt = 0, .name = MakeOptional(CharSpan(tzStore.name, sizeof(tzStore.name))) };
    }
}
```

Note that the empty span is made to be `sizeof(tzStore.name)` rather than zero. Maybe we should take-out the strlen stuff here, and simply fix that...

Reply:
I agree that fixing the initialisation part is the better approach. That is a leftover mistake from changing the other parts.

Reply:
It requires more fixes. The bug also occurs at:

```
 576   │     if (lastTzState != TimeState::kInvalid)
 577   │     {
 578   │         const auto & tzStore = GetTimeZone()[0];
 579   │         lastTz.offset        = tzStore.timeZone.offset;
 580   │         if (tzStore.timeZone.name.HasValue())
 581   │         {
 582   │             lastTz.name.SetValue(CharSpan(name));   /// HERE - this is sizeof as well
 583   │             memcpy(name, tzStore.name, sizeof(tzStore.name));
 584   │         }
 585   │     }
 ```
 
 And maybe other places. I did not try to chase down all instances, because in the existing code it looks like it assumes "name" is updated and may not actually use CharSpan.

Reply:
@fessehaeve - would you like to take the time to fix this properly? I have limited bandwidth for this. My opinion is that we should not have had this extra container at all. Since this is only used during serialization, the correct thing to do is to only instantiate it there. Having the same data stored twice (once string, once span and keeping in sync) is a foot-gun.

Reply:
Loading for broken data at line 614 is also not going to work:

```
614   │         if (newTz.name.HasValue() && newTz.name.Value().size() > 0)
 615   │         {
 616   │             size_t len = newTz.name.Value().size(); // This is 64 because of before ...
 617   │             if (len > sizeof(tzStore.name))
 618   │             {
 619   │                 ReturnErrorOnFailure(LoadTimeZone());
 620   │                 return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_IB;
 621   │             }
 622   │             memset(tzStore.name, 0, sizeof(tzStore.name));
 623   │             chip::MutableCharSpan tempSpan(tzStore.name, len);
 624   │             if (CHIP_NO_ERROR != CopyCharSpanToMutableCharSpan(newTz.name.Value(), tempSpan))
 625   │             {
 626   │                 ReturnErrorOnFailure(LoadTimeZone());
 627   │                 return CHIP_IM_GLOBAL_STATUS(InvalidCommand);
 628   │             }
 629   │             tzStore.timeZone.name.SetValue(CharSpan(tzStore.name, len));
 630   │         }
 631   │         else
 632   │         {
 633   │             tzStore.timeZone.name.ClearValue();
 634   │         }
 ```
 
 It would have worked if we had right sizes to start with, but existing devices will still load 0-fill values.

Reply:
This indeed needs fixing in a more detailed way. I don't have bandwidth before end of next week because of the coming up Nov 20 deadline.

- I don't understand how the span is supposed to work without the backing buffer.
- Why is line 616 broken? Isn't `newTz` supposed to have the correct size?

Reply:
for line 616, because name.Value is a CharSpan, I expect it is serialized and deserialized with embedded nulls.

So if an old device did `tzStore.timeZone = { .offset = 0, .validAt = 0, .name = MakeOptional(CharSpan(tzStore.name, sizeof(tzStore.name))) };` it created a 64-byte zero-fill charspan.

So this code will not correct old data serialized with embedded nulls and will preserve them. This is assuming that some devices have been prodiced and they save and load timezones. 

So if `newTz.name.Value()` would not have embedded nulls, it would be fine. However I claim because of this bug already existing, this newTz does have embedded nulls.

Given bandwidth constraints, are you ok for this to go in as is and then you can followup with the actual correct fix?

Reply:
This ideally should be fixed by Nov20 - it is a spec violation (and makes our subscriptions slightly slower because it sends an extra 64 null bytes for no reason)

Reply:
I understand now.
I can only commit to this after Nov 20.